# CPSC 330 - Applied Machine Learning 

## Homework 5: Evaluation metrics
### Associated lectures: [Lectures 9, 10](https://ubc-cs.github.io/cpsc330/README.html) 

**Due date: Monday, Feb 27, 2023 at 11:59pm**

## Imports

In [1]:
import os
import re
import sys
from hashlib import sha1

import matplotlib.pyplot as plt
import numpy as np
import pandas as pd
# import tests_hw5
from sklearn import datasets
from sklearn.compose import make_column_transformer
from sklearn.dummy import DummyClassifier, DummyRegressor
from sklearn.ensemble import RandomForestClassifier, RandomForestRegressor
from sklearn.linear_model import LogisticRegression, Ridge
from sklearn.metrics import (
    accuracy_score,
    classification_report,
    confusion_matrix,
    f1_score,
    make_scorer,
    precision_score,
    recall_score,
)
from sklearn.model_selection import (
    GridSearchCV,
    RandomizedSearchCV,
    cross_val_score,
    cross_validate,
    train_test_split,
)
from sklearn.pipeline import Pipeline, make_pipeline
from sklearn.preprocessing import OneHotEncoder, OrdinalEncoder, StandardScaler

## Instructions 
<hr>
rubric={points:3}

Follow the [homework submission instructions](https://github.com/UBC-CS/cpsc330-2022W2/blob/main/docs/homework_instructions.md). 

**You may work with a partner on this homework and submit your assignment as a group.** Below are some instructions on working as a group.  
- The maximum group size is 2. 
- Use group work as an opportunity to collaborate and learn new things from each other. 
- Be respectful to each other and make sure you understand all the concepts in the assignment well. 
- It's your responsibility to make sure that the assignment is submitted by one of the group members before the deadline. 
- You can find the instructions on how to do group submission on Gradescope [here](https://help.gradescope.com/article/m5qz2xsnjy-student-add-group-members).

<br><br>

## Exercise 1: Precision, recall, and f1 score by hand <a name="1"></a>
<hr>

Consider the problem of predicting whether a patient has cancer or not. It is important to catch this disease early to reduce mortality rate; late diagnosis will result in metastasis to other organs, which adversely impacts patient's prognosis. Below are confusion matrices of two machine learning models: Model A and Model B. 

- Model A

|         | Predicted disease | Predicted no disease |
| :------------- | -----------------------: | -----------------------: |
| **Actual disease**       | 48 | 32 |
| **Actual no disease**       | 20 | 100 |


- Model B

|        | Predicted disease | Predicted no disease |
| :------------- | -----------------------: | -----------------------: |
| **Actual disease**       | 43 | 22 |
| **Actual no disease**       | 35 | 100 |

### 1.1 Positive vs. negative class 
rubric={points:2}

**Your tasks:**

Precision, recall, and f1 score depend upon which class is considered "positive", that is the thing you wish to find. In the example above, which class is likely to be the "positive" class? Why? 

Answer:  
"disease" is likely to be the "positive" class, because that is what our model wishes to find.

<br><br>

### 1.2 Accuracy
rubric={points:2}

**Your tasks:**

Calculate accuracies for Model A and Model B. 

We'll store all metrics associated with Model A and Model B in the `results_dict` below. 

In [2]:
results_dict = {"A": {}, "B": {}}

In [3]:
results_dict["A"]["accuracy"] = (48+100)/(48+100+20+32)
results_dict["B"]["accuracy"] = (43+100)/(43+100+22+35)
print(results_dict["A"]["accuracy"])
print(results_dict["B"]["accuracy"])

0.74
0.715


<br><br>

### 1.3 Which model would you pick? 
rubric={points:1}

**Your tasks:**

Which model would you pick simply based on the accuracy metric? 

Answer:  
Model A, since its accuracy is higher.

<br><br>

### 1.4 Precision, recall, f1-score
rubric={points:6}

**Your tasks:**

1. Calculate precision, recall, f1-score for Model A and Model B manually, without using `scikit-learn` tools. 


In [4]:
A_prec = results_dict["A"]["precision"] = 48/68
B_prec = results_dict["B"]["precision"] = 43/(43+35)
A_recall = results_dict["A"]["recall"] = 48/80
B_recall = results_dict["B"]["recall"] = 43/(43+22)
results_dict["A"]["f1"] = (2*A_prec*A_recall)/(A_prec+A_recall)
results_dict["B"]["f1"] = (2*B_prec*B_recall)/(B_prec+B_recall)

Show the dataframe with all results. 

In [5]:
pd.DataFrame(results_dict)

A         B
accuracy   0.740000  0.715000
precision  0.705882  0.551282
recall     0.600000  0.661538
f1         0.648649  0.601399

<br><br>

### 1.5 Discussion
rubric={points:4}

**Your tasks:**
1. Given the type of problem (early cancer diagnosis), which metric is more informative in this problem? Why? 
2. Which model would you pick based on this information? 

Answer:  
1. Recall is more informative, as we do not want to falsely diagnose anyone who actually has cancer as not having cancer. In other words, we want to minimize the number of false negatives in our model and recall is more informative in telling us that.  
2. I would pick Model B as it has a higher recall than Model A.

<br><br>

### (Optional) 1.6 
rubric={points:1}

**Your tasks:**

Provide 2 to 3 example classification datasets (with links) where accuracy metric would be misleading. Discuss which evaluation metric would be more appropriate for each dataset. You may consider datasets we have used in this course so far. You could also look up datasets on Kaggle. 

<br><br><br><br>

### Exercise 2: Classification evaluation metrics using `sklearn` <a name="2"></a>
<hr>

In general, when a dataset is imbalanced, accuracy does not provide the whole story. In class, we looked at credit card fraud dataset which is a classic example of an imbalanced dataset. 

Another example is customer churn datasets. [Customer churn](https://en.wikipedia.org/wiki/Customer_attrition) refers to the notion of customers leaving a subscription service like Netflix. In this exercise, we will try to predict customer churn in a dataset where most of the customers stay with the service and a small minority cancel their subscription. To start, please download the [Kaggle telecom customer churn dataset](https://www.kaggle.com/becksddf/churn-in-telecoms-dataset). Once you have the data, you should be able to run the following code:

The starter code below reads the data CSV as a pandas dataframe and splits it into 70% train and 30% test. 

Note that `churn` column in the dataset is the target. "True" means the customer left the subscription (churned) and "False" means they stayed.

> Note that for this kind of problem a more appropriate technique is something called survival analysis and we'll be talking about it later in the course. For now, we'll just treat it as a binary classification problem. 

In [6]:
df = pd.read_csv("bigml_59c28831336c6604c800002a.csv", encoding="utf-8")
train_df, test_df = train_test_split(df, test_size=0.3, random_state=123)
train_df

state  account length  area code phone number international plan  \
1402    NE              70        415     421-8535                 no   
1855    WI              67        510     417-2265                 no   
633     NJ             122        415     327-9341                 no   
1483    NV             107        510     419-9688                yes   
2638    HI             105        510     364-8128                 no   
...    ...             ...        ...          ...                ...   
2154    WY             126        408     339-9798                yes   
3089    WV              70        510     348-3777                 no   
1766    NJ             125        415     406-6400                 no   
1122    NE             159        415     362-5111                 no   
1346    PA             106        408     403-9167                yes   

     voice mail plan  number vmail messages  total day minutes  \
1402              no                      0              213.4   
1855              no                      0              109.1   
633              yes                     34              146.4   
1483              no                      0              234.1   
2638              no                      0              125.4   
...              ...                    ...                ...   
2154              no                      0              197.6   
3089             yes                     30              143.4   
1766              no                      0              182.3   
1122              no                      0              189.1   
1346              no                      0              133.7   

      total day calls  total day charge  ...  total eve calls  \
1402               86             36.28  ...               77   
1855              134             18.55  ...               76   
633               104             24.89  ...              103   
1483               91             39.80  ...              105   
2638              116             21.32  ...               95   
...               ...               ...  ...              ...   
2154              126             33.59  ...              112   
3089               72             24.38  ...               92   
1766               64             30.99  ...              121   
1122              105             32.15  ...              147   
1346               45             22.73  ...              107   

      total eve charge  total night minutes  total night calls  \
1402             17.40                256.6                101   
1855             12.10                 91.2                 86   
633               7.62                220.0                 91   
1483             13.86                282.5                100   
2638             22.23                241.6                104   
...                ...                  ...                ...   
2154             20.95                285.3                104   
3089             14.45                127.9                 68   
1766             11.88                171.6                 96   
1122             20.92                242.0                106   
1346             15.96                181.9                 89   

      total night charge  total intl minutes  total intl calls  \
1402               11.55                 5.7                 4   
1855                4.10                10.9                 5   
633                 9.90                15.6                 4   
1483               12.71                10.0                 3   
2638               10.87                11.4                 9   
...                  ...                 ...               ...   
2154               12.84                12.5                 8   
3089                5.76                 9.4                 4   
1766                7.72                11.6                 7   
1122               10.89                10.4                 5   
1346                8.19           

In [7]:
# train_df.describe()
train_df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 2333 entries, 1402 to 1346
Data columns (total 21 columns):
 #   Column                  Non-Null Count  Dtype  
---  ------                  --------------  -----  
 0   state                   2333 non-null   object 
 1   account length          2333 non-null   int64  
 2   area code               2333 non-null   int64  
 3   phone number            2333 non-null   object 
 4   international plan      2333 non-null   object 
 5   voice mail plan         2333 non-null   object 
 6   number vmail messages   2333 non-null   int64  
 7   total day minutes       2333 non-null   float64
 8   total day calls         2333 non-null   int64  
 9   total day charge        2333 non-null   float64
 10  total eve minutes       2333 non-null   float64
 11  total eve calls         2333 non-null   int64  
 12  total eve charge        2333 non-null   float64
 13  total night minutes     2333 non-null   float64
 14  total night calls       2333 non-null

<br><br>

### 2.1 Distribution of target values
rubric={points:4}

**Your tasks:**

Examine the distribution of target values in the train split. Do you see class imbalance? If yes, do we need to deal with it? Why or why not? 

In [8]:
print(train_df["churn"].value_counts(normalize=True))

False    0.850407
True     0.149593
Name: churn, dtype: float64


Answer:  
Yes, there is a class imbalance. We need to deal with it because a class imbalance will misrepresent our accuracy. This is especially so because we want to identify examples that have `True` values for "churn", but this is the class that is the minority in our dataset.

<br><br>

### (Optional) 2.2 EDA 
rubric={points:1}

**Your tasks:**

Come up with **two** exploratory questions you would like to answer and explore those. Briefly discuss your results in 1-3 sentences.

You are welcome to use `pandas_profiling` (see Lecture 10) but you don't have to.

<br><br>

### 2.3 Column transformer 
rubric={points:14}

The code below creates `X_train`, `y_train`, `X_test`, `y_test` for you. 
In preparation for building a classifier, set up a `ColumnTransformer` that performs whatever feature transformations you deem sensible. This can include dropping features if you think they are not helpful. Remember that by default `ColumnTransformer` will drop any columns that aren't accounted for when it's created.

For each group of features (e.g. numeric, categorical or else) explain why you are applying the particular transformation. For example, "I am doing transformation X to the following categorical features: `a`, `b`, `c` because of reason Y," etc.

Finally, fit `ColumnTransformer` on your training set; and use the `ColumnTransformer` to transform your train data.

In [9]:
X_train = train_df.drop(columns=["churn"])
X_test = test_df.drop(columns=["churn"])

y_train = train_df["churn"]
y_test = test_df["churn"]

numeric_feats = ["account length", "number vmail messages", "total day minutes", "total day calls", "total day charge", 
                "total eve minutes", "total eve calls", "total eve charge", "total night minutes", "total night calls", 
                 "total night charge", "total intl minutes", "total intl calls", "total intl charge", "customer service calls"]
# We apply Standard Scaling to these numeric features as Logistic Regression is a gradient-based algorithm
# and is sensitive to the range of the data points

categorical_feats = ["state", "area code", "international plan", "voice mail plan"]

# We apply One-Hot encoding on categorical features which are not ordinal. 
# This is so that they are converted into their numeric representatiomn and hence
# can be used by the model.

passthrough_feats = []

drop_feats = ["phone number"]

ct = make_column_transformer(
    (StandardScaler(), numeric_feats),  # scaling on numeric features
    (OneHotEncoder(drop="if_binary", sparse=False), categorical_feats),  # OHE on categorical features
    ("passthrough", passthrough_feats),  # no transformations on the binary features
    ("drop", drop_feats),  # drop the drop features
)

<br><br>

<br><br>

### 2.4 area code feature
rubric={points:4}

The original dataset had a feature called `area code`.

1. The area codes are numbers. Does it make sense to encode them as one-hot-endoded (OHE) or not? Please justify your response.
2. What were the possible values of `area code`? 
3. If area code is encoded with OHE, how many new features are created to replace it?

In [10]:
train_df["area code"].value_counts()

415    1178
408     588
510     567
Name: area code, dtype: int64

1. Yes because it only includes 3 types of values. More importantly, for a feature like `area code`, the values in these feature do not have any meaningful quantitative qualities.  
2. The possible values for `area code` are `415`, `408` and `510`
3. Three new features are created to replace it.

<br><br>

### 2.5 Logistic regression
rubric={points:12} 

**Your tasks:**

1. Report the cross-validation results of a `LogisticRegression` model, with default Hparams, on the following metrics: `"accuracy", "precision", "recall", "f1"`
2. Are you satisfied with the results? Explain why or why not. Discuss in a few sentences. 

In [11]:
lr = LogisticRegression()
pipe = make_pipeline(ct, lr)
pipe.fit(X_train, y_train)

scoring = [
    "accuracy",
    "precision",
    "recall",
    "f1",    
]  # scoring can be a string, a list, or a dictionary
scores = cross_validate(
    pipe, X_train, y_train, return_train_score=True, scoring=scoring
)
pd.DataFrame(scores)

fit_time  score_time  test_accuracy  train_accuracy  test_precision  \
0  0.040106    0.009091       0.869379        0.864416        0.666667   
1  0.051091    0.010210       0.852248        0.868167        0.520000   
2  0.043965    0.009588       0.850107        0.867095        0.500000   
3  0.039690    0.008971       0.869099        0.863953        0.642857   
4  0.037022    0.008940       0.839056        0.868773        0.413793   

   train_precision  test_recall  train_recall   test_f1  train_f1  
0         0.630000     0.257143      0.225806  0.371134  0.332454  
1         0.654206     0.185714      0.250896  0.273684  0.362694  
2         0.639640     0.171429      0.254480  0.255319  0.364103  
3         0.620370     0.260870      0.239286  0.371134  0.345361  
4         0.651786     0.171429      0.261649  0.242424  0.373402

<br><br>

### 2.6 Logistic regression with `class_weight`
rubric={points:6}

**Your tasks:**

1. Set the `class_weight` parameter of your logistic regression model to `'balanced'` and report the same metrics as in the previous part. 
2. Do you prefer this model to the one in the previous part? Discuss your results in a few sentences while comparing the metrics of this model and the previous model.

In [18]:
lr = LogisticRegression(class_weight="balanced")
pipe = make_pipeline(ct, lr)
pipe.fit(X_train, y_train)

scoring = [
    "accuracy",
    "precision",
    "recall",
    "f1",    
]  # scoring can be a string, a list, or a dictionary
scores = cross_validate(
    pipe, X_train, y_train, return_train_score=True, scoring=scoring
)
pd.DataFrame(scores)

fit_time  score_time  test_accuracy  train_accuracy  test_precision  \
0  0.054868    0.012829       0.785867        0.769561        0.380952   
1  0.046030    0.008652       0.768737        0.771168        0.366197   
2  0.035922    0.008734       0.764454        0.774384        0.348485   
3  0.035021    0.008902       0.751073        0.779325        0.340136   
4  0.053037    0.008575       0.733906        0.786824        0.320000   

   train_precision  test_recall  train_recall   test_f1  train_f1  
0         0.369151     0.685714      0.763441  0.489796  0.497664  
1         0.372852     0.742857      0.777778  0.490566  0.504065  
2         0.378007     0.657143      0.788530  0.455446  0.511034  
3         0.385017     0.724638      0.789286  0.462963  0.517564  
4         0.395797     0.685714      0.810036  0.436364  0.531765

We prefer this model over the model where the classes were not balanced as the `test_f1` is higher and has less variance across each of the cross-validation folds.

<br><br>

### 2.7 Hyperparameter optimization
rubric={points:10}

1. Jointly optimize `C` and `class_weight` with `GridSearchCV` and `scoring="f1"`.
  - For `class_weight`, consider 3 values: 
    - `None` (no weight)
    - "weight of class 0 = 1"  and  "weight of class 1 = 3"
    - '`balanced`'
  - For `C`, choose some reasonable values
2. What values of `C` and `class_weight` are chosen and what is the best cross-validation f1 score?

In [24]:
pipe = make_pipeline(
    ct, 
    LogisticRegression(max_iter=1000)
)

param_grid = {
    "logisticregression__C": 10.0 ** np.arange(-2, 2, 0.5), 
    "logisticregression__class_weight": [None, {0:10000, 1:3}, 'balanced'], 
}

grid_search = GridSearchCV(
    pipe, param_grid, cv=5, n_jobs=-1, return_train_score=True, scoring="f1"
)

grid_search.fit(X_train, y_train)

best_score = grid_search.best_score_
print(best_score)

best_class_weight = grid_search.best_params_['logisticregression__class_weight']
best_C = grid_search.best_params_['logisticregression__C']
print("Best class_weight = ", best_class_weight)
print("Best C = ", best_C)

0.47873599874436196
Best class_weight =  balanced
Best C =  0.1



<br><br>

### 2.8 Test results
rubric={points:10}

**Your tasks**
1. Evaluate the best model on the test set. In particular show each of the following on the test set:  
    - Plot Confusion matrix
    - Plot Precision-recall curve 
    - Calculate average precision score
    - Plot ROC curve
    - Report AUC score
3. Comment on the AUC score and give an intuitive explanation of what this value of AUC means for this problem.

<br><br><br><br>

### Exercise 3: Regression metrics <a name="3"></a>
<hr> 


For this exercise, we'll use [California housing dataset](https://scikit-learn.org/stable/modules/generated/sklearn.datasets.fetch_california_housing.html) from `sklearn datasets`. The code below loads the dataset.  

In [14]:
from sklearn.datasets import fetch_california_housing

housing_df = fetch_california_housing(as_frame=True).frame

### 3.1: Data spitting and exploration 
rubric={points:4}

**Your tasks:**

1. Split the data into train (75%) and test (25%) splits. 
2. Explore the train split. Do you need to apply any transformations on the data? If yes, create a preprocessor with the appropriate transformations. 
3. Separate `X` and `y` to train and test splits. 

<br><br>

### 3.2 Baseline: Linear Regression 
rubric={points:2}

**Your tasks:**
1. Carry out cross-validation using `sklearn.linear_model.LinearRegression` with default scoring. 
2. What metric is used for scoring by default? 

<br><br>

### 3.3 Random Forest Regressor
rubric={points:7}

In this exercise, we are going to use [`RandomForestRegressor`](https://scikit-learn.org/stable/modules/generated/sklearn.ensemble.RandomForestRegressor.html) model which we haven't looked into yet. At this point you should feel comfortable using models with our usual ML workflow even if you don't know the details. We'll talk about `RandomForestRegressor` later in the course.  

The code below defines a custom scorer called `mape_scorer` and creates dictionaries for two model (`models`) and five evaluation metrics (`score_types_reg`). 

**Your tasks:**

1. Using the `models` and the evaluation metrics `score_types_reg` in the code below, carry out cross-validation with each model, by passing the evaluation metrics to `scoring` argument of `cross_validate`. Use a pipeline with the model as an estimator if you are applying any transformations. 
2. Show results as a dataframe. 
3. Interpret the results. How do the models compare to the baseline? Which model seems to be performing well with different metrics? 


In [15]:
models = {
    "Ridge": Ridge(),
    "Random Forest": RandomForestRegressor(),
}

score_types_reg = {
    "neg_mean_squared_error": "neg_mean_squared_error",
    "neg_root_mean_squared_error": "neg_root_mean_squared_error",
    "neg_mean_absolute_error": "neg_mean_absolute_error",
    "r2": "r2",
    "neg_mean_absolute_percentage_error": "neg_mean_absolute_percentage_error",
}

<br><br>

### 3.4 Hyperparameter optimization 
rubric={points:1}

1. Carry out hyperparameter optimization using `RandomizedSearchCV` and `Ridge` with the following `param_dist`. The `alpha` hyperparameter of `Ridge` controls the fundamental tradeoff. Choose `neg_mean_absolute_percentage_error` as the HParam optimization metric.

2. What was the best `alpha` hyper-parameter found?

In [16]:
from scipy.stats import loguniform

param_dist = {"ridge__alpha": loguniform(1e-3, 1e3)}

<br><br>

### 3.5 Test results
rubric={points:4}

**Your tasks:**

Test the best model (from 3.4) on the test set based on the `neg_mean_absolute_percentage_error` score.

<br><br>

### 3.6 Model interpretation  
rubric={points:4}

Ridge is a linear model and it learns coefficients associated with each feature during `fit()`. 

**Your tasks:**

1. Explore coefficients learned by the `Ridge` model above as a pandas dataframe with two columns: 
   - features 
   - coefficients
2. Increasing which feature values would result in higher housing price? 

<br><br>

## Submission instructions 

**PLEASE READ:** When you are ready to submit your assignment do the following:

1. Run all cells in your notebook to make sure there are no errors by doing `Kernel -> Restart Kernel and Clear All Outputs` and then `Run -> Run All Cells`. 
2. Notebooks with cell execution numbers out of order or not starting from “1” will have marks deducted. Notebooks without the output displayed may not be graded at all (because we need to see the output in order to grade your work).
3. Upload the assignment using Gradescope's drag and drop tool. Check out this [Gradescope Student Guide](https://lthub.ubc.ca/guides/gradescope-student-guide/) if you need help with Gradescope submission. 